pkg/epp/datalayer/metrics/client.go

```diff
@@ -0,0 +1,135 @@
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datalayer"
+)
+
+// Client is an interface for retrieving the metrics from an endpoint URL.
+type Client interface {
+	Get(ctx context.Context, target *url.URL, ep datalayer.Addressable) (PrometheusMetricMap, error)
+}
+
+// ClientFactory returns a Client suitable for an endpoint.
+// Implementations may return a new Client each time, or use a cached
+// copy for optimized retrieval.
+type ClientFactory interface {
+	GetClientForEndpoint(ep datalayer.Addressable) Client
+}
+
+// GetDefaultClientFactory returns a default implementation of the
+// ClientFactory, which caches and reuses client across calls.
+func GetDefaultClientFactory() ClientFactory {
+	return defaultClientFactory
+}
+
+// -- package implementations --
+var (
+	cleanupTick          = 30 * time.Second
+	maxIdleTime          = time.Minute
+	defaultClientFactory = newClientFactory()
+)
+
+type client struct {
+	cl       *http.Client
+	lastUsed time.Time
+}
+
+type clientmap struct {
+	cleanupOnce sync.Once
+	clients     sync.Map // map[string]*client (uses client IP as key)
+}
+
+func newClientFactory() *clientmap {
+	clm := &clientmap{}
+	clm.startCleanupGoroutine()
+	return clm
+}
+
+func (clm *clientmap) startCleanupGoroutine() {
+	clm.cleanupOnce.Do(func() {
+		go func() {
+			for {
+				time.Sleep(cleanupTick)
+				now := time.Now()
+
+				clm.clients.Range(func(key, value any) bool {
+					entry := value.(*client)
+					if now.Sub(entry.lastUsed) > maxIdleTime {
+						clm.clients.Delete(key)
+					}
+					return true
+				})
+			}
+		}()
+	})
+}
+
+func (clm *clientmap) GetClientForEndpoint(ep datalayer.Addressable) Client {
+	cl := newClient()
+	id := ep.GetIPAddress()
+
+	if value, loaded := clm.clients.LoadOrStore(id, cl); !loaded {
+		cl = value.(*client) // let it panic if need be
+	}
+	cl.lastUsed = time.Now()
+	return cl
+}
+
+func newClient() *client {
+	return &client{
+		cl: &http.Client{
+			Timeout: 10 * time.Second,
+			// TODO: set additional timeouts, transport options, etc.
+		},
+		lastUsed: time.Now(),
+	}
+}
+
+func (cl *client) Get(ctx context.Context, target *url.URL, ep datalayer.Addressable) (PrometheusMetricMap, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, target.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %v", err)
+	}
+	resp, err := cl.cl.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch metrics from %s: %w", ep.GetNamespacedName(), err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code from %s: %v", ep.GetNamespacedName(), resp.StatusCode)
+	}
+
+	parser := expfmt.TextParser{}
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return metricFamilies, err
```

Comment:
metricFamilies is prometheus representation of the data, right?
do you expect every extractor to parse it to an easier format?
what are your thoughts about converting this into (key, value) pairs and allow extractor to extract a subset of (key,value) pairs?
so the conversion from metricFamilies to (key,value) pairs is done only once?

Reply:
The code runs in the Collector, not every extractor. You can look at the extractor's expected input type.
Can make this easier to see by adding an output type to the Collector's interface (the only reason would be "document" the output type - the code uses it when an Extractor is registered so caller access to output type is not really needed).

Reply:
> I can create some helper functions that can be imported and reused, but let's do that when we write the first non-standard extractor? So we have concrete knowledge of what to reuse and how to expose it. Makes sense?

yes 👍